<a href="https://colab.research.google.com/github/Induwara427/Statistical-Learning-e21427/blob/main/E21427_Assignment_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***Question 3***

1)
The number of transformer failures in one day can be:

0,1,2,3.

Thus the sample space is

Ω={0,1,2,3}.

Since the space is finite, we take the sigma-algebra to be the power set:

F=P(Ω).

The probability measure P is defined by

P({0})=0.50,

P({1})=0.30,

P({2})=0.15,

P({3})=0.05.

The probabilities sum to 1:

0.50+0.30+0.15+0.05=1.

Therefore,

(Ω,F,P)=({0,1,2,3},P(Ω),P)


2)
Define

X:Ω→R

by

X(ω)=ω.

Thus:

X=0 means no transformer failures,

X=1 means one failure,

X=2 means two failures,

X=3 means three failures.

3)
The probability mass function is

p
X
	​
(x)=P(X=x).

Hence,

p
X
	​
(x)=
	​

0.50,
x=0

0.30,
x=1

0.15,
x=2

0.05,
x=3

0,

​otherwise.

4)
The expectation of a discrete random variable is

E[X]=
x
∑
	​
xP(X=x).

Thus,

E[X]=0(0.50)+1(0.30)+2(0.15)+3(0.05).

Compute:

=0+0.30+0.30+0.15

=0.75.

5)
First compute

E[X^
2
].
E[X
2
]=0^
2
(0.50)+1^
2
(0.30)+2^
2
(0.15)+3^
2
(0.05).

=0+0.30+0.60+0.45

=1.35.

Variance formula:

Var(X)=E[X^
2
]−(E[X])^
2
.

Thus,

Var(X)=1.35−(0.75)
2
.

=1.35−0.5625

=0.7875.

6)
(a) P(X≥1)

P(X≥1)=P(X=1)+P(X=2)+P(X=3)

=0.30+0.15+0.05

=0.50.


(b) P(X≥2)

P(X≥2)=P(X=2)+P(X=3)

=0.15+0.05

=0.20. ​

7)
Expectation E[X]

The expectation represents the average number of emergency transformer failures per day.

Since

E[X]=0.75,

the district experiences, on average,

0.75 transformer failures per day
	​


over a long period.

Variance Var(X)

The variance measures how much the number of failures fluctuates from day to day.

A larger variance means:

failures are less predictable

daily counts vary more widely.

Here,

Var(X)=0.7875
	​



***Question 4***

1)


In [1]:
# Constructing the probability space (Ω, F, P)
# for X ~ Beta(2,18)

# Sample space Ω
Omega = "[0,1]"

# Sigma-algebra F
F = "Borel sigma-algebra on [0,1]"

# Beta parameters
alpha = 2
beta = 18

# Define Beta function using math library
import math

def beta_function(a, b):
    return (math.gamma(a) * math.gamma(b)) / math.gamma(a + b)

B = beta_function(alpha, beta)

# Display probability measure definition
print("Probability Space (Ω, F, P)\n")

print("Ω =", Omega)
print("F =", F)

print("\nBeta parameters:")
print("alpha =", alpha)
print("beta =", beta)

print("\nBeta function B(2,18) =", B)

print("\nProbability density function:")
print("f(x) = (1/B(2,18)) * x^(1) * (1-x)^(17),  0 < x < 1")

Probability Space (Ω, F, P)

Ω = [0,1]
F = Borel sigma-algebra on [0,1]

Beta parameters:
alpha = 2
beta = 18

Beta function B(2,18) = 0.0029239766081871343

Probability density function:
f(x) = (1/B(2,18)) * x^(1) * (1-x)^(17),  0 < x < 1


2)
Define

X:Ω→R

by

X(ω)=ω.

Thus X records the proportion of defective items in the batch.

Since X takes values continuously in [0,1], it is a continuous random variable.

3)


In [2]:
# Probability Density Function of X
# X ~ Beta(2,18)

import sympy as sp

# Define symbols
x = sp.symbols('x', positive=True)

# Parameters
alpha = 2
beta = 18

# Define Beta function
B = sp.beta(alpha, beta)

# Define the PDF
f_x = (1 / B) * (x**(alpha - 1)) * ((1 - x)**(beta - 1))

# Simplify expression
f_x_simplified = sp.simplify(f_x)

# Print the PDF
print("Probability Density Function of X:\n")
sp.pprint(f_x_simplified)

print("\nFor 0 < x < 1")

Probability Density Function of X:

          17 
-x⋅(x - 1)   
─────────────
  Β(2, 18)   

For 0 < x < 1


4)


In [3]:
# Verify that f(x) is a valid probability density function
# for X ~ Beta(2,18)

import sympy as sp

# Define variable
x = sp.symbols('x', real=True)

# Parameters
alpha = 2
beta = 18

# Define Beta function
B = sp.beta(alpha, beta)

# Define the density function
f = (1 / B) * (x**(alpha - 1)) * ((1 - x)**(beta - 1))

# Step 1: Check non-negativity
print("Density function f(x):")
sp.pprint(f)

print("\nFor 0 < x < 1, f(x) >= 0")

# Step 2: Verify total integral equals 1
integral = sp.integrate(f, (x, 0, 1))

print("\nIntegral of f(x) from 0 to 1:")
sp.pprint(sp.simplify(integral))

# Final verification
if sp.simplify(integral) == 1:
    print("\nVerified: f(x) is a valid probability density function.")
else:
    print("\nVerification failed.")

Density function f(x):
         17
x⋅(1 - x)  
───────────
 Β(2, 18)  

For 0 < x < 1, f(x) >= 0

Integral of f(x) from 0 to 1:
     1      
────────────
342⋅Β(2, 18)

Verification failed.


5)
For a Beta distribution:

E[X]=
α/α+β
	​

Here,

α=2,

β=18.

Thus,

E[X]=
2/20
	​

=0.10

6)


In [4]:
# Compute the variance Var(X)
# for X ~ Beta(2,18)

import sympy as sp

# Parameters
alpha = 2
beta = 18

# Variance formula for Beta distribution
variance = (alpha * beta) / (((alpha + beta)**2) * (alpha + beta + 1))

# Simplify result
variance_simplified = sp.simplify(variance)

# Print result
print("Variance Var(X):")
sp.pprint(variance_simplified)

print("\nDecimal value:")
print(float(variance_simplified))

Variance Var(X):
0.00428571428571429

Decimal value:
0.004285714285714286


7)


In [5]:
# Compute P(X <= 0.10)
# for X ~ Beta(2,18)

import sympy as sp

# Define variable
x = sp.symbols('x', real=True)

# Parameters
alpha = 2
beta = 18

# Define Beta function
B = sp.beta(alpha, beta)

# Define the PDF
f = (1 / B) * (x**(alpha - 1)) * ((1 - x)**(beta - 1))

# Compute probability P(X <= 0.10)
probability = sp.integrate(f, (x, 0, 0.10))

# Simplify result
probability_simplified = sp.N(probability)

# Print results
print("P(X <= 0.10) =")
sp.pprint(probability_simplified)

P(X <= 0.10) =
0.579735021168403


8)


In [6]:
# Compute P(0.05 <= X <= 0.15)
# for X ~ Beta(2,18)

import sympy as sp

# Define variable
x = sp.symbols('x', real=True)

# Parameters
alpha = 2
beta = 18

# Define Beta function
B = sp.beta(alpha, beta)

# Define the PDF
f = (1 / B) * (x**(alpha - 1)) * ((1 - x)**(beta - 1))

# Compute probability
probability = sp.integrate(f, (x, 0.05, 0.15))

# Numerical approximation
probability_numeric = sp.N(probability)

# Print result
print("P(0.05 <= X <= 0.15) =")
sp.pprint(probability_numeric)

P(0.05 <= X <= 0.15) =
0.556215488790056


9)
The Beta distribution is suitable because:

it is defined on the interval [0,1],

proportions and probabilities naturally lie between 0 and 1,

it can model many different shapes depending on parameters.

For α=2 and β=18:

the distribution is concentrated near small values,

which reflects a manufacturing process with generally low defect rates.

Thus the Beta distribution is a realistic model for production quality proportions.

10)
Expectation

E[X]=0.10

means the average defective proportion is about:

10%
	​


over many production batches.

Variance

Var(X)≈0.00429

measures how much the defect proportion fluctuates between batches.

A relatively small variance indicates:

production quality is fairly stable,

defect proportions do not vary dramatically from day to day.

***Question 5***

1)
The support of a Triangular distribution is:

[a,b].

Thus,

4≤X≤10

	​
2)


In [7]:
# Probability Density Function of X
# X ~ Triangular(a=4, b=10, c=6)

import sympy as sp

# Define variable
x = sp.symbols('x', real=True)

# Parameters
a = 4
b = 10
c = 6

# Define the piecewise PDF
f_x = sp.Piecewise(
    ((2 * (x - a)) / ((b - a) * (c - a)), (x >= a) & (x <= c)),
    ((2 * (b - x)) / ((b - a) * (b - c)), (x > c) & (x <= b)),
    (0, True)
)

# Simplify the function
f_x_simplified = sp.simplify(f_x)

# Print the PDF
print("Probability Density Function f_X(x):\n")
sp.pprint(f_x_simplified)

Probability Density Function f_X(x):

⎧x   2                     
⎪─ - ─   for x ≥ 4 ∧ x ≤ 6 
⎪6   3                     
⎪                          
⎨5   x                     
⎪─ - ──  for x ≤ 10 ∧ x > 6
⎪6   12                    
⎪                          
⎩  0         otherwise     


3)


In [11]:
import sympy as sp

# Define variable
x = sp.Symbol('x')

# Define the piecewise density function
f1 = 6 / (x - 4)          # for 4 <= x <= 6
f2 = 12 / (10 - x)        # for 6 <= x <= 10

# Compute integrals on each interval
I1 = sp.integrate(f1, (x, 4, 6))
I2 = sp.integrate(f2, (x, 6, 10))

# Total probability
total = I1 + I2

print("Integral from 4 to 6:", I1)
print("Integral from 6 to 10:", I2)
print("Total integral:", total)

# Verification
print("Is total equal to 1?", sp.simplify(total) == 1)

Integral from 4 to 6: oo
Integral from 6 to 10: oo + 12*I*pi
Total integral: oo + 12*I*pi
Is total equal to 1? False


4)
4. Compute the Expectation E[X]

For a Triangular distribution:

E[X]=
a+b+c/3
	​

Thus,

E[X]=
4+10+6/3
	​
=
20/3
	​

Final Answer
E[X]=
20/3
	​

≈6.67 hours

	​
5)
5. Compute the Variance Var(X)

Formula:

Var(X)=
a^
2
+b^
2
+c^
2
−ab−ac−bc/
18
	​


Substitute values:

=
16+100+36−40−24−60/18
	​

=
28/
18
	​

=
14/
9
	​

Final Answer

Var(X)=
14/9
	​

≈1.56



6)

a=4:

shortest possible installation time.

b=10:

longest possible installation time.

c=6:

most likely installation time.

These values summarize expert knowledge about the project duration.

7)
The Triangular distribution is useful when:

exact historical data is limited,

but experts can estimate:

minimum,

maximum,

most likely values.

It is widely used in:

project management,

construction planning,

risk analysis,

PERT-style estimation.

Thus it is appropriate for estimating installation duration.


8)


In [14]:
import sympy as sp

# Define variable
x = sp.Symbol('x')

# Parameters
a, b, c = 4, 10, 6

# Left-side PDF (since 5 is in [4,6])
f = (x - a) / ((b - a) * (c - a))

# Compute probability P(X ≤ 5)
prob = sp.integrate(f, (x, 4, 5))

print("P(X ≤ 5):", prob)
print("Decimal:", float(prob))

P(X ≤ 5): 1/24
Decimal: 0.041666666666666664


9)


In [15]:
import sympy as sp

# Define variable
x = sp.Symbol('x')

# Parameters
a, b, c = 4, 10, 6

# PDF pieces
f_left = (x - a) / ((b - a) * (c - a))     # 4 to 6
f_right = (b - x) / ((b - a) * (b - c))    # 6 to 10

# Compute split integrals
part1 = sp.integrate(f_left, (x, 5, 6))
part2 = sp.integrate(f_right, (x, 6, 8))

# Total probability
prob = part1 + part2

print("P(5 ≤ X ≤ 8):", prob)
print("Decimal:", float(prob))

P(5 ≤ X ≤ 8): 3/8
Decimal: 0.375


10)
This model helps planners:

estimate expected project duration,

evaluate schedule uncertainty,

compute risk of delays,

allocate labor/resources effectively.


In sensitivity analysis:

changing a, b, or c shows how uncertainty affects timelines,

helping managers identify critical risks and bottlenecks.

Thus the model supports better project scheduling and decision-making.

***Question 6***

1)


In [16]:
# Parameters
a = 8
b = 20
c = 11

# Compute shape parameters
alpha = 1 + 4 * ((c - a) / (b - a))
beta  = 1 + 4 * ((b - c) / (b - a))

print("alpha =", alpha)
print("beta =", beta)

alpha = 2.0
beta = 4.0


2)
X∈[8,20]

3)


In [17]:
import numpy as np
from scipy.special import beta

# Parameters
a = 8
b = 20
c = 11

# Shape parameters
alpha = 1 + 4 * ((c - a) / (b - a))
beta_param = 1 + 4 * ((b - c) / (b - a))

# PERT PDF function
def pert_pdf(x):
    if a <= x <= b:
        numerator = (x - a)**(alpha - 1) * (b - x)**(beta_param - 1)
        denominator = beta(alpha, beta_param) * (b - a)**(alpha + beta_param - 1)
        return numerator / denominator
    else:
        return 0

# Example evaluation
x_values = np.linspace(a, b, 5)
pdf_values = [pert_pdf(x) for x in x_values]

for x, f in zip(x_values, pdf_values):
    print(f"x = {x:.2f}, f(x) = {f:.6f}")

x = 8.00, f(x) = 0.000000
x = 11.00, f(x) = 0.175781
x = 14.00, f(x) = 0.104167
x = 17.00, f(x) = 0.019531
x = 20.00, f(x) = 0.000000


4)
A standard continuous model:

(Ω,F,P)

Ω=[8,20]

F = Borel σ-algebra on [8,20]

P defined by:

P(A)=∫
A
	​
f(x)dx

5)
X:Ω→[8,20], X(ω)=ω

6)


In [18]:
import numpy as np
from scipy.integrate import quad
from scipy.special import beta

# Parameters
a = 8
b = 20
c = 11

# Shape parameters
alpha = 1 + 4 * ((c - a) / (b - a))
beta_param = 1 + 4 * ((b - c) / (b - a))

# PERT PDF
def pert_pdf(x):
    if a <= x <= b:
        numerator = (x - a)**(alpha - 1) * (b - x)**(beta_param - 1)
        denominator = beta(alpha, beta_param) * (b - a)**(alpha + beta_param - 1)
        return numerator / denominator
    else:
        return 0

# 1. Theoretical expectation
E_theoretical = (a + 4*c + b) / 6

# 2. Numerical expectation via integration
E_numerical, error = quad(lambda x: x * pert_pdf(x), a, b)

print("Theoretical E[X]:", E_theoretical)
print("Numerical E[X]:", E_numerical)
print("Integration error estimate:", error)

Theoretical E[X]: 12.0
Numerical E[X]: 12.0
Integration error estimate: 1.3322676295501878e-13


7)


In [19]:
import math

# Parameters
a = 8
b = 20
c = 11

# Step 1: Compute expectation
E = (a + 4*c + b) / 6

# Step 2: Compute variance approximation
variance = ((E - a) * (b - E)) / 7

# Step 3: Standard deviation (optional but useful)
std_dev = math.sqrt(variance)

print("E[X] =", E)
print("Variance =", variance)
print("Standard deviation =", std_dev)

E[X] = 12.0
Variance = 4.571428571428571
Standard deviation = 2.138089935299395


8)


In [20]:
import math

# Parameters
a = 8
b = 20
c = 11

# Step 1: Expected value
E = (a + 4*c + b) / 6

# Step 2: Variance approximation
variance = ((E - a) * (b - E)) / 7

# Step 3: Standard deviation
std_dev = math.sqrt(variance)

print("E[X] =", E)
print("Variance =", variance)
print("Standard Deviation =", std_dev)

E[X] = 12.0
Variance = 4.571428571428571
Standard Deviation = 2.138089935299395


9)
Triangular distribution assumes straight-line increase/decrease → sharper, less realistic tails.

PERT distribution:

smoother (Beta-based)

places more probability around the most likely value c

reduces extreme tail weight (less pessimistic/optimistic bias)

In this case:

Triangular mean:

8+11+20/
3
	​

=13

PERT mean:

12

10)
a=8: best-case (fastest completion)

b=20: worst-case (delays, risks)

c=11: most likely duration (expert estimate peak)

E[X]=12: realistic planning baseline (expected duration)

σ≈2.14: uncertainty level (moderate variability around 12 days)

***Question 7***



In [21]:
import math

# Given values
a = 5
c = 11
E = 11  # expected value equals most likely time

# 1. Solve for b using E = (a + 4c + b) / 6
b = 6 * E - a - 4 * c

# 2. Compute variance using PERT approximation
variance = ((E - a) * (b - E)) / 7

# 3. Standard deviation (optional but useful)
std_dev = math.sqrt(variance)

print("b =", b)
print("Variance =", variance)
print("Standard deviation =", std_dev)

b = 17
Variance = 5.142857142857143
Standard deviation = 2.2677868380553634
